conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java

```diff
@@ -291,22 +293,33 @@ private CodeBlock endpointInvocation(EndpointDefinition endpointDefinition, List
                 code.addStatement("$1T $2N = $3N.bodySerDe().deserializeInputStream($4N)",
                         InputStream.class, bodyParam.getArgName().get(), RUNTIME_VAR_NAME, EXCHANGE_VAR_NAME);
             } else {
-                code.addStatement("$1T $2N = $3N.deserialize($4N)",
-                        typeMapper.getClassName(bodyParam.getType()).box(),
-                        bodyParam.getArgName().get(),
-                        DESERIALIZER_VAR_NAME,
-                        EXCHANGE_VAR_NAME);
+                // This class uses a strict type mapper to force immutable wrapper types.
+                // Where our service interface may expect List<String>, we deserialize to ImmutableList<String>
+                // for behavior consistent with generated beans, and pass the value to the service.
+                // This becomes problematic whe inner types are expanded. For example Optional<List<String>>
+                // becomes Optional<ImmutableList<String>>, and cannot be cast to the service type.
+                // In this case, we cast all the way down to object, then back to our target type to
+                // avoid compilation errors without sacrificing immutability.
+                TypeName handlerType = mappers.handlerRequest().getClassName(bodyParam.getType()).box();
+                TypeName serviceType = mappers.request().getClassName(bodyParam.getType()).box();
+                boolean requiresCast = requiresForcefulCast(handlerType, serviceType);
+                if (requiresCast) {
+                    code.add("@$1T($2S)", SuppressWarnings.class, "unchecked");
+                }
+                code.addStatement(requiresCast
+                                ? "$1T $2N = ($1T) (Object) $3N.deserialize($4N)" : "$1T $2N = $3N.deserialize($4N)",
+                        serviceType, bodyParam.getArgName().get(), DESERIALIZER_VAR_NAME, EXCHANGE_VAR_NAME);
```

Comment:
This piece is more complicated than I'd like, however I'd rather take this complication in a conjure generator over debugging problems caused by mutable state.